### PR TITLE
Verify with DS

### PR DESCRIPTION
**Pull request by khibino**
Add logic to verify DNSKEY with DS.

implemented digest algorithm list

* SHA1/SHA256/SHA384

### CODE REVIEW

File: dnsext-dnssec/DNS/SEC/Verify/Types.hs [hunk 1]
```diff
@@ -43,3 +44,20 @@ verifyRRSIGwith RRSIGImpl{..} RD_DNSKEY{..} rrsig@RD_RRSIG{..} rr = do
...
+  forall digest .
+  DSImpl
+  { dsIGetDigest :: ByteString -> digest
+  , dsIVerify :: digest -> ByteString -> Bool
+  }
+
+verifyDSwith :: DSImpl -> Domain -> RD_DNSKEY -> RD_DS -> Either String ()
+verifyDSwith DSImpl{..} owner dnskey@RD_DNSKEY{..} RD_DS{..} = do
+  unless (dnskey_pubalg == ds_pubalg) $
+    Left $ "verifyDS: pubkey algorithm mismatch between DNSKEY and DS: " ++ show dnskey_pubalg ++ " =/= " ++ show ds_pubalg
```
Comment (kazu-yamamoto) [inline dnsext-dnssec/DNS/SEC/Verify/Types.hs:58]: g/"verifyDS"/"verifyDSwith"/

File: dnsext-dnssec/DNS/SEC/Verify/Types.hs [hunk 2]
```diff
@@ -43,3 +44,20 @@ verifyRRSIGwith RRSIGImpl{..} RD_DNSKEY{..} rrsig@RD_RRSIG{..} rr = do
...
+  {- TODO: check DNSKEY with keytag -}
+  let digest = dsIGetDigest $ runSPut $ putDomain Canonical owner >> putResourceData Canonical dnskey
+      ds_digest' = Opaque.toByteString ds_digest
+  unless (dsIVerify digest ds_digest') $
+    Left "verifyDS: rejected on verification"
```
Comment (kazu-yamamoto) [inline dnsext-dnssec/DNS/SEC/Verify/Types.hs:63]: ditto

**Review by kazu-yamamoto — CHANGES_REQUESTED** [review]
Please change minor things.

File: dnsext-dnssec/DNS/SEC/Verify/Types.hs (hunk 1 above)
Comment (khibino) [reply dnsext-dnssec/DNS/SEC/Verify/Types.hs]: https://github.com/kazu-yamamoto/dnsext/pull/26/commits/d03206c608280f414f60a7d5f12c64eb5c9b2415

**Review by kazu-yamamoto — APPROVED** [review]
LGTM

**Comment by kazu-yamamoto** [comment]
Merged.

**Pull request merged**